accord-core/src/main/java/accord/utils/SimpleBitSet.java

```diff
@@ -0,0 +1,348 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package accord.utils;
+
+import java.util.Arrays;
+
+import static java.lang.Long.highestOneBit;
+import static java.lang.Long.lowestOneBit;
+import static java.lang.Long.numberOfTrailingZeros;
+
+public class SimpleBitSet
+{
+    public static class SerializationSupport
+    {
+        public static long[] getArray(SimpleBitSet bs)
+        {
+            return bs.bits;
+        }
+
+        public static SimpleBitSet construct(long[] bits)
+        {
+            return new SimpleBitSet(bits);
+        }
+    }
+
+    final long[] bits;
+    int count;
+
+    public SimpleBitSet(int size)
+    {
+        bits = new long[(size + 63)/64];
+    }
+
+    public SimpleBitSet(int size, boolean set)
+    {
+        this(size);
+        if (set)
+        {
+            Arrays.fill(bits, 0, size / 64, -1L);
+            if ((size & 63) != 0)
+                bits[indexOf(size - 1)] = -1L >>> (64 - (size & 63));
+            count = size;
+        }
+    }
+
+    public SimpleBitSet(SimpleBitSet copy)
+    {
+        bits = copy.bits.clone();
+        count = copy.count;
+    }
+
+    SimpleBitSet(long[] bits)
+    {
+        this.bits = bits;
+        for (long v : bits)
+            count += Long.bitCount(v);
+    }
+
+    public boolean set(int i)
+    {
+        int index = indexOf(i);
+        long bit = bit(i);
+        if (0 != (bits[index] & bit))
+            return false;
+        bits[index] |= bit;
+        ++count;
+        return true;
+    }
+
+    public void setRange(int from, int to)
+    {
+        if (to <= from)
+        {
+            Invariants.checkArgument(to >= from, "to < from (%s < %s)", to, from);
```

Comment:
if `to <= from` then we check `to >= from`... which means we require `to == from`? 

```suggestion
            Invariants.checkArgument(to == from, "to < from (%s < %s)", to, from);
```

Reply:
The following test fails, 

```
@Test
    public void setRange()
    {
        int size = 64 * 8;
        SimpleBitSet set = new SimpleBitSet(size);
        set.setRange(10, 10);
        Assertions.assertThat(set.get(10)).isTrue();
    }
```

when `from == to` we no-op, *should* we reject this case and simplify the condition?

Reply:
That is likely to just complicate callers? It is typical in utility classes only to reject _invalid_ parameters, not ones that are valid but no-op?

Reply:
I am ok with no-op, we should just fix the check to be more clear.  If we don't no-op the single caller `this.waitingOnCommit.setRange(0, deps.keyDeps.txnIdCount());` would need to be careful for range only deps

Reply:
I don’t **mind** the change, but I disagree it’s clearer - any `>=` would be a bug, and locally that’s all we care about. We don’t currently perform the check unless we’ve already tested `<`, so you can pre-apply this and reduce the test to `=`. But that’s not clearer in my book, it just relies on the prior statement to apply the full condition. The modified check is also prone to being disarmed by refactoring. So I think it is strictly worse, but minimally so.

Reply:
> any >= would be a bug

you point out that `==` is a no-op and supported case, so `>=` being a bug isn't true, its only true in the context of the if check above, which makes you have to undo the logic in your head (making it more complex)...


The following is simpler and very clear IMO

```
Invariants.checkArgument(from <= to, "to < from (%s < %s)", to, from);
```

> We don’t currently perform the check unless we’ve already tested <, so you can pre-apply this and reduce the test to =

2 checks when we can perform in 1?

> The modified check is also prone to being disarmed by refactoring. So I think it is strictly worse, but minimally so.

I don't follow this argument... we need a if check that adds more complication to this check become a refactor in the future might change things?

Reply:
Why is `from <= to` clearer than `to >= from`? This is identical to the original check!

> 2 checks when we can perform in 1?

I was discussing your proposal, which seems to suggest we convert the inequality check into an equality check - which is preapplying the restriction of the if condition in which the test is applied. But there is no value in applying this reduction, is precisely my point.

Reply:
> Why is from <= to simpler than to >= from? This is identical to the original check!

Sorry if I am not clear, I was asking we change

```
if (to <= from)
{
    Invariants.checkArgument(to >= from, "to < from (%s < %s)", to, from);
}
```

to

```
Invariants.checkArgument(from <= to, "to < from (%s < %s)", to, from);
```

the reason its simpler is that we have 1 single check rather than 2, which requires the reads to undo to figure out the actual single logical check being done.  With the previous logic we want to fail in the following boolean condition: `(to <= from) && !(to >= from)`. Which simplifies `(to <= from) && (to < from)` which simplifies again to `to < from`... so with this `checkArgument` api that means we want `to >= from`